File: 2022/lang/nim/src/solution/day06.nim
from std/sequtils import deduplicate

proc shiftMarker(arr: array[4, char], insert: char): array[4, char] =
  for a in 1..3:
    result[a-1] = arr[a]
  result[3] = insert

proc endOfMarkerIndex(): int =
  let input = open("input/day-06.txt")
  defer: input.close()
  var
    iterations = 4
    potentialMarker: array[4, char]
    cursor: array[1, char]
  discard readChars(input, potentialMarker)
  while readChars(input, cursor) > 0:
    if deduplicate(potentialMarker).len() == 4:
      break
    potentialMarker = shiftMarker(potentialMarker, cursor[0])
    iterations += 1
  return iterations

proc partOne() =
  echo("  - Part 1: ", endOfMarkerIndex())

proc solve*() =
  echo("- Day 06")
  partOne()

File: 2022/lang/rust/src/solution/day_06.rs
use std::{
    collections::{BTreeSet, VecDeque},
    fs::File,
    io::{self, BufRead, BufReader},
};

pub fn solve() -> io::Result<()> {
    println!("- Day 06");
    let file = File::open("input/day-06.txt")?;
    part_one(BufReader::new(file));
    Ok(())
}

fn part_one(reader: BufReader<File>) {
    let result: usize = reader
        .lines()
        .filter_map(|l| l.ok())
        .map(|l| {
            let mut i = 4;
            let mut window = l.chars().take(i).collect::<VecDeque<char>>();
            let mut rest = l.chars().skip(i);
            while !is_uniq(&window) {
                window.push_back(rest.next().unwrap());
                window.pop_front();
                i += 1;
            }
            i
        })
        .sum();
    println!("  - Part one: {result}");
}

fn is_uniq(v: &VecDeque<char>) -> bool {
    BTreeSet::<&char>::from_iter(v.iter()).len() == v.len()
}
